File: src/lib/webhook-templates.ts
/**
 * Shared webhook payload templates.
 *
 * Both the Dashboard test webhook and the Python bot render these JSON
 * templates with mustache-style `{{variable}}` substitution before POSTing to
 * Discord. Admins edit them via /dashboard/admin/webhooks.
 *
 * Substitution is string-based and happens BEFORE JSON.parse, so:
 *  - String values must be quoted: `"value": "{{keyword}}"`
 *  - Numeric values unquoted:      `"color": {{color}}`
 *  - All string vars are JSON-escaped via `jsonEscape()` to prevent broken JSON
 *    when product titles or keywords contain `"`, `\`, newlines, etc.
 */

// ─── Defaults (match the current hardcoded bot payloads 1:1) ─────────────────

export const DEFAULT_USER_WEBHOOK_TEMPLATE = `{
  "username": "Eventry",
  "embeds": [
    {
      "title": "\u{1F680} Autostart Triggered",
      "color": 11382189,
      "timestamp": "{{timestamp_iso}}",
      "fields": [
        { "name": "Keyword",      "value": "\`{{keyword}}\`",                              "inline": true  },
        { "name": "Status",       "value": "{{status_emoji}} {{status}}",                  "inline": true  },
        { "name": "Product",      "value": "{{product_title}}",                            "inline": false },
        { "name": "Price Breaks", "value": "{{price_info}}",                               "inline": true  },
        { "name": "Stock",        "value": "{{stock_info}}",                               "inline": true  },
        { "name": "Product Link", "value": "{{product_link}}",                             "inline": false },
        { "name": "Message",      "value": "[Jump to original message]({{message_jump_url}})", "inline": false }
      ],
      "footer": { "text": "Eventry Autostart \u2022 {{timestamp_cest}}" }
    }
  ]
}`

export const DEFAULT_ADMIN_WEBHOOK_TEMPLATE = `{
  "username": "Eventry Admin Log",
  "embeds": [
    {
      "title": "\u{1F514} Autostart Triggered",
      "color": {{color}},
      "timestamp": "{{timestamp_iso}}",
      "fields": [
        { "name": "Product",        "value": "{{product_title}}",                            "inline": false },
        { "name": "Product Link",   "value": "{{product_link}}",                             "inline": false },
        { "name": "Status",         "value": "{{status_emoji}} {{status_summary}}",          "inline": true  },
        { "name": "Channel",        "value": "#{{channel_name}}",                            "inline": true  },
        { "name": "Source Message", "value": "[Jump to original]({{message_jump_url}})",     "inline": false },
        { "name": "Users ({{user_count}})", "value": "{{user_list}}",                        "inline": false }
      ],
      "footer": { "text": "Eventry Admin Log \u2022 {{timestamp_cest}}" }
    }
  ]
}`

// ─── Variable catalogs (shown in the editor UI) ──────────────────────────────

export interface TemplateVariable {
  name: string
  description: string
}

export const USER_TEMPLATE_VARIABLES: TemplateVariable[] = [
  { name: 'keyword',           description: 'Matched keyword text' },
  { name: 'status',            description: '"200 OK" or "500 Failed"' },
  { name: 'status_emoji',      description: '✅ or ❌' },
  { name: 'http_status',       description: 'Raw HTTP status code (number as string)' },
  { name: 'product_title',     description: 'Product title from Silently embed' },
  { name: 'product_description', description: 'Product description from embed' },
  { name: 'price_info',        description: 'Price breaks string' },
  { name: 'stock_info',        description: 'Stock string' },
  { name: 'product_link',      description: 'Decoded product URL (or —)' },
  { name: 'message_jump_url',  description: 'Discord jump URL of the source message' },
  { name: 'timestamp_iso',     description: 'ISO-8601 UTC timestamp' },
  { name: 'timestamp_cest',    description: 'DD.MM.YYYY HH:MM:SS CEST' },
]

/**
 * Admin log webhook is AGGREGATED — ONE webhook per quicktask event,
 * containing every user who was triggered. Variables are event-level only.
 */
export const ADMIN_TEMPLATE_VARIABLES: TemplateVariable[] = [
  { name: 'product_title',    description: 'Product title from Silently embed' },
  { name: 'product_link',     description: 'Decoded product URL (or —)' },
  { name: 'channel_name',     description: 'Discord channel name' },
  { name: 'message_jump_url', description: 'Discord jump URL of the source message' },
  { name: 'user_list',        description: 'Multiline list: <@discord_id> · `silently_key` · `keyword` · status' },
  { name: 'user_count',       description: 'Total number of users triggered' },
  { name: 'success_count',    description: 'Users with http_status == 200' },
  { name: 'failure_count',    description: 'Users with failure status' },
  { name: 'status_summary',   description: 'e.g. "5 OK / 2 Failed"' },
  { name: 'status_emoji',     description: '✅ all OK / ⚠️ mixed / ❌ all failed' },
  { name: 'color',            description: 'Decimal color (green / amber / red)' },
  { name: 'timestamp_iso',    description: 'ISO-8601 UTC timestamp' },
  { name: 'timestamp_cest',   description: 'DD.MM.YYYY HH:MM:SS CEST' },
]

// ─── Helpers ─────────────────────────────────────────────────────────────────

/**
 * JSON-escape a string so it can safely be inserted as a template variable
 * INSIDE a JSON string context (e.g. `"value": "{{keyword}}"`).
 *
 * `JSON.stringify(s).slice(1, -1)` handles \", \\, \n, \t, \r and control
 * characters exactly the way JSON.parse will accept them.
 */
export function jsonEscape(s: string): string {
  return JSON.stringify(s).slice(1, -1)
}

/**
 * Substitute all {{name}} tokens in `template` with the matching value in
 * `vars`, then JSON.parse the result. Missing variables are replaced with an
 * empty string. Returns `null` on parse errors so callers can fall back.
 *
 * String vars should already be jsonEscape()'d by the caller; numeric vars
 * can be passed as numbers and will be inserted without quotes.
 */
export function renderWebhookTemplate(
  template: string,
  vars: Record<string, string | number>,
): Record<string, unknown> | null {
  const rendered = template.replace(/\{\{(\w+)\}\}/g, (_, name: string) => {
    const v = vars[name]
    if (v === undefined || v === null) return ''
    return String(v)
  })
  try {
    const parsed = JSON.parse(rendered)
    if (typeof parsed !== 'object' || parsed === null || Array.isArray(parsed)) return null
    return parsed as Record<string, unknown>
  } catch {
    return null
  }
}

/**
 * Build the standard vars dict for a USER webhook render. All string inputs
 * are JSON-escaped automatically.
 */
export function buildUserWebhookVars(input: {
  keyword: string
  http_status: number
  product_title: string
  product_description: string
  price_info: string
  stock_info: string
  product_link: string
  message_jump_url: string
}): Record<string, string | number> {
  const now = new Date()
  const cest = formatCest(now)
  const ok = input.http_status === 200
  return {
    keyword:             jsonEscape(input.keyword),
    status:              jsonEscape(`${input.http_status} ${ok ? 'OK' : 'Failed'}`),
    status_emoji:        ok ? '\u2705' : '\u274C',
    http_status:         String(input.http_status),
    product_title:       jsonEscape(input.product_title.slice(0, 256)),
    product_description: jsonEscape(input.product_description),
    price_info:          jsonEscape(input.price_info.slice(0, 256)),
    stock_info:          jsonEscape(input.stock_info.slice(0, 64)),
    product_link:        jsonEscape(input.product_link),
    message_jump_url:    jsonEscape(input.message_jump_url),
    timestamp_iso:       now.toISOString(),
    timestamp_cest:      jsonEscape(cest),
  }
}

export interface AdminLogUserResult {
  discord_user_id: string | null
  whop_user_id: string
  silently_key: string
  keyword: string
  http_status: number
}

/**
 * Build the vars dict for the aggregated admin log webhook render.
 *
 * Unlike user webhooks (one per user), the admin log fires ONCE per autostart
 * event and lists every participating user inline. Vars are event-level.
 */
export function buildAdminWebhookVars(input: {
  product_title: string
  product_link: string
  channel_name: string
  message_jump_url: string
  results: AdminLogUserResult[]
}): Record<string, string | number> {
  const now = new Date()
  const cest = formatCest(now)
  const { results } = input

  const userCount = results.length
  const successCount = results.filter(r => r.http_status === 200).length
  const failureCount = userCount - successCount

  let statusEmoji: string
  let color: number
  if (userCount === 0) {
    statusEmoji = '\u2753' // ❓
    color = 8421504 // gray
  } else if (failureCount === 0) {
    statusEmoji = '\u2705' // ✅
    color = 4906624 // green
  } else if (successCount === 0) {
    statusEmoji = '\u274C' // ❌
    color = 16287345 // red
  } else {
    statusEmoji = '\u26A0\uFE0F' // ⚠️
    color = 16753920 // amber
  }
  const statusSummary = `${successCount} OK / ${failureCount} Failed`

  // Build user_list lines: <@discord> · `silently_key` · `keyword` · status
  const lines = results.map(r => {
    const mention = r.discord_user_id ? `<@${r.discord_user_id}>` : '`' + r.whop_user_id + '`'
    const ok = r.http_status === 200
    const lineStatus = ok ? `\u2705 ${r.http_status} OK` : `\u274C ${r.http_status} Failed`
    return `${mention} \u00B7 \`${r.silently_key || '—'}\` \u00B7 \`${r.keyword || '—'}\` \u00B7 ${lineStatus}`
  })
  let userList = lines.join('\n')
  if (userList.length > 1000) {
    const truncated: string[] = []
    let running = 0
    for (const ln of lines) {
      if (running + ln.length + 1 > 950) break
      truncated.push(ln)
      running += ln.length + 1
    }
    truncated.push(`\u2026 and ${userCount - truncated.length} more`)
    userList = truncated.join('\n')
  }

  return {
    product_title:    jsonEscape(input.product_title.slice(0, 256)),
    product_link:     jsonEscape(input.product_link),
    channel_name:     jsonEscape(input.channel_name || '—'),
    message_jump_url: jsonEscape(input.message_jump_url),
    user_list:        jsonEscape(userList),
    user_count:       String(userCount),
    success_count:    String(successCount),
    failure_count:    String(failureCount),
    status_summary:   jsonEscape(statusSummary),
    status_emoji:     statusEmoji,
    color,
    timestamp_iso:    now.toISOString(),
    timestamp_cest:   jsonEscape(cest),
  }
}

// ─── Sample data for the "Send Test" button ──────────────────────────────────

export const SAMPLE_USER_INPUT = {
  keyword: 'test-keyword',
  http_status: 200,
  product_title: 'Test Product — Webhook Verification',
  product_description: 'Sample description for the test webhook',
  price_info: '85.00: 9, 95.00: 6',
  stock_info: '15',
  product_link: 'https://example.com/product/test',
  message_jump_url: 'https://discord.com',
}

export const SAMPLE_ADMIN_INPUT: {
  product_title: string
  product_link: string
  channel_name: string
  message_jump_url: string
  results: AdminLogUserResult[]
} = {
  product_title: 'Test Product — Webhook Verification',
  product_link: 'https://example.com/product/test',
  channel_name: 'test-channel',
  message_jump_url: 'https://discord.com',
  results: [
    {
      discord_user_id: '123456789012345678',
      whop_user_id: 'user_abc123',
      silently_key: 'silently_key_1_DEMO',
      keyword: 'test-keyword',
      http_status: 200,
    },
    {
      discord_user_id: '987654321098765432',
      whop_user_id: 'user_def456',
      silently_key: 'silently_key_2_DEMO',
      keyword: 'test-keyword',
      http_status: 200,
    },
    {
      discord_user_id: null,
      whop_user_id: 'user_ghi789',
      silently_key: 'silently_key_3_DEMO',
      keyword: 'another-keyword',
      http_status: 500,
    },
  ],
}

// ─── Private helpers ─────────────────────────────────────────────────────────

function formatCest(d: Date): string {
  // DD.MM.YYYY HH:MM:SS CEST  (matches Python bot format)
  const parts = new Intl.DateTimeFormat('de-DE', {
    timeZone: 'Europe/Berlin',
    day: '2-digit',
    month: '2-digit',
    year: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: false,
  }).formatToParts(d)
  const get = (t: string) => parts.find(p => p.type === t)?.value ?? ''
  return `${get('day')}.${get('month')}.${get('year')} ${get('hour')}:${get('minute')}:${get('second')} CEST`
}

File: bot/services/webhook_templates.py
"""
Shared Discord webhook payload templates for the Python bot.

Templates are edited in the Dashboard Admin Panel and stored in the
`app_settings` table under:
    - webhook_user_payload_template
    - webhook_admin_payload_template

Substitution uses mustache-style `{{variable}}` tokens and happens BEFORE
json.loads, so string vars must be JSON-escaped via `json_escape()` first.

These defaults 1:1 mirror the previous hardcoded payloads in webhooks.py.
"""

import json
import logging
import re
from datetime import datetime, timezone
from typing import Optional

import pytz

logger = logging.getLogger(__name__)


DEFAULT_USER_WEBHOOK_TEMPLATE = r"""{
  "username": "Eventry",
  "embeds": [
    {
      "title": "\ud83d\ude80 Autostart Triggered",
      "color": 11382189,
      "timestamp": "{{timestamp_iso}}",
      "fields": [
        { "name": "Keyword",      "value": "`{{keyword}}`",                              "inline": true  },
        { "name": "Status",       "value": "{{status_emoji}} {{status}}",                "inline": true  },
        { "name": "Product",      "value": "{{product_title}}",                          "inline": false },
        { "name": "Price Breaks", "value": "{{price_info}}",                             "inline": true  },
        { "name": "Stock",        "value": "{{stock_info}}",                             "inline": true  },
        { "name": "Product Link", "value": "{{product_link}}",                           "inline": false },
        { "name": "Message",      "value": "[Jump to original message]({{message_jump_url}})", "inline": false }
      ],
      "footer": { "text": "Eventry Autostart \u2022 {{timestamp_cest}}" }
    }
  ]
}"""


DEFAULT_ADMIN_WEBHOOK_TEMPLATE = r"""{
  "username": "Eventry Admin Log",
  "embeds": [
    {
      "title": "\ud83d\udd14 Autostart Triggered",
      "color": {{color}},
      "timestamp": "{{timestamp_iso}}",
      "fields": [
        { "name": "Product",        "value": "{{product_title}}",                        "inline": false },
        { "name": "Product Link",   "value": "{{product_link}}",                         "inline": false },
        { "name": "Status",         "value": "{{status_emoji}} {{status_summary}}",      "inline": true  },
        { "name": "Channel",        "value": "#{{channel_name}}",                        "inline": true  },
        { "name": "Source Message", "value": "[Jump to original]({{message_jump_url}})", "inline": false },
        { "name": "Users ({{user_count}})", "value": "{{user_list}}",                    "inline": false }
      ],
      "footer": { "text": "Eventry Admin Log \u2022 {{timestamp_cest}}" }
    }
  ]
}"""


_PLACEHOLDER_RE = re.compile(r"\{\{(\w+)\}\}")


def json_escape(s: str) -> str:
    """
    Escape a string so it can be safely inlined inside a JSON string literal.
    Produces exactly what json.loads will accept — handles ", \\, \n, \t, \r,
    and control characters. Mirrors JS `JSON.stringify(s).slice(1, -1)`.
    """
    if s is None:
        return ""
    dumped = json.dumps(s, ensure_ascii=False)
    return dumped[1:-1]


def render_template(template: str, variables: dict) -> Optional[dict]:
    """
    Substitute `{{name}}` tokens in `template`, then `json.loads` the result.
    Returns `None` on parse error so callers can fall back to a default.
    """
    def _sub(m: re.Match) -> str:
        name = m.group(1)
        val = variables.get(name)
        if val is None:
            return ""
        return str(val)

    rendered = _PLACEHOLDER_RE.sub(_sub, template)
    try:
        parsed = json.loads(rendered)
    except json.JSONDecodeError as e:
        logger.warning(f"[WEBHOOK TEMPLATE] JSON parse error: {e} | rendered[:200]={rendered[:200]!r}")
        return None
    if not isinstance(parsed, dict):
        logger.warning("[WEBHOOK TEMPLATE] Rendered payload is not an object")
        return None
    return parsed


def build_user_vars(
    *,
    keyword: str,
    http_status: int,
    product_title: str,
    product_description: str,
    price_info: str,
    stock_info: str,
    product_link: str,
    message_jump_url: str,
) -> dict:
    """Build the variable dict for the user autostart webhook template."""
    now_utc = datetime.now(timezone.utc)
    now_cest = datetime.now(pytz.timezone("Europe/Berlin"))
    cest_str = now_cest.strftime("%d.%m.%Y %H:%M:%S CEST")
    ok = http_status == 200
    return {
        "keyword":             json_escape(keyword),
        "status":              json_escape(f"{http_status} {'OK' if ok else 'Failed'}"),
        "status_emoji":        "\u2705" if ok else "\u274c",
        "http_status":         str(http_status),
        "product_title":       json_escape(product_title[:256]),
        "product_description": json_escape(product_description or ""),
        "price_info":          json_escape(price_info[:256]),
        "stock_info":          json_escape(stock_info[:64]),
        "product_link":        json_escape(product_link),
        "message_jump_url":    json_escape(message_jump_url),
        "timestamp_iso":       now_utc.isoformat(),
        "timestamp_cest":      json_escape(cest_str),
    }


def build_admin_vars(
    *,
    product_title: str,
    product_link: str,
    channel_name: str,
    message_jump_url: str,
    results: list[dict],
) -> dict:
    """
    Build the variable dict for the admin log webhook template.

    `results` is a list of dicts — one per user who participated in the same
    quicktask autostart event. Each dict must contain:
        { discord_user_id, whop_user_id, silently_key, keyword, http_status }

    The resulting webhook is aggregated into ONE payload (not per user).
    """
    now_utc = datetime.now(timezone.utc)
    now_cest = datetime.now(pytz.timezone("Europe/Berlin"))
    cest_str = now_cest.strftime("%d.%m.%Y %H:%M:%S CEST")

    user_count = len(results)
    success_count = sum(1 for r in results if r.get("http_status") == 200)
    failure_count = user_count - success_count

    if user_count == 0:
        status_emoji = "\u2753"  # ❓
        color = 8421504  # gray
    elif failure_count == 0:
        status_emoji = "\u2705"  # ✅
        color = 4906624          # green
    elif success_count == 0:
        status_emoji = "\u274c"  # ❌
        color = 16287345         # red
    else:
        status_emoji = "\u26a0\ufe0f"  # ⚠️
        color = 16753920               # amber

    status_summary = f"{success_count} OK / {failure_count} Failed"

    # Build a multi-line user list: `<@discord>` · `silently_key` · `keyword` · status
    lines: list[str] = []
    for r in results:
        dc_id = r.get("discord_user_id") or ""
        whop = r.get("whop_user_id") or ""
        mention = f"<@{dc_id}>" if dc_id else f"`{whop}`"
        key = r.get("silently_key") or "—"
        keyword = r.get("keyword") or "—"
        http = r.get("http_status", 0)
        ok = http == 200
        line_status = f"\u2705 {http} OK" if ok else f"\u274c {http} Failed"
        lines.append(f"{mention} \u00b7 `{key}` \u00b7 `{keyword}` \u00b7 {line_status}")

    # Cap to 1024 chars (Discord field value limit) with graceful truncation
    user_list = "\n".join(lines)
    if len(user_list) > 1000:
        truncated_lines: list[str] = []
        running = 0
        for ln in lines:
            if running + len(ln) + 1 > 950:
                break
            truncated_lines.append(ln)
            running += len(ln) + 1
        remaining = user_count - len(truncated_lines)
        truncated_lines.append(f"\u2026 and {remaining} more")
        user_list = "\n".join(truncated_lines)

    return {
        "product_title":   json_escape(product_title[:256]),
        "product_link":    json_escape(product_link),
        "channel_name":    json_escape(channel_name or "—"),
        "message_jump_url": json_escape(message_jump_url),
        "user_list":       json_escape(user_list),
        "user_count":      str(user_count),
        "success_count":   str(success_count),
        "failure_count":   str(failure_count),
        "status_summary":  json_escape(status_summary),
        "status_emoji":    status_emoji,
        "color":           color,
        "timestamp_iso":   now_utc.isoformat(),
        "timestamp_cest":  json_escape(cest_str),
    }
